File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/argument/PairArgument.java
package net.volcano.jdacommands.model.command.arguments.implementation.argument;

import kotlin.Pair;
import lombok.experimental.SuperBuilder;
import net.volcano.jdacommands.exceptions.command.parsing.InvalidArgumentsException;
import net.volcano.jdacommands.model.command.arguments.CommandArgument;
import net.volcano.jdacommands.model.command.arguments.implementation.ArgumentParsingData;

@SuperBuilder
public class PairArgument extends CommandArgument<Pair<?, ?>> {
	
	protected final CommandArgument<?> typeArgument1;
	protected final CommandArgument<?> typeArgument2;
	
	@Override
	public Pair<?, ?> parseValue(ArgumentParsingData data) throws InvalidArgumentsException {
		
		try {
			return new Pair<>(typeArgument1.parseValue(data), null);
		} catch (InvalidArgumentsException e) {
			return new Pair<>(null, typeArgument2.parseValue(data));
		}
		
	}
	
	@Override
	public String getUsage() {
		return "<" + typeArgument1.getUsage() + " | " + typeArgument2 + ">";
	}
}

File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/argument/PairArgument.kt
package net.volcano.jdacommands.model.command.arguments.implementation.argument

import net.volcano.jdacommands.exceptions.command.parsing.InvalidArgumentsException
import net.volcano.jdacommands.model.command.arguments.CommandArgument
import net.volcano.jdacommands.model.command.arguments.implementation.ArgumentParsingData

class PairArgument(
	private val typeArgument1: CommandArgument<*>,
	private val typeArgument2: CommandArgument<*>
) : CommandArgument<Pair<Any?, Any?>?>() {

	@Throws(InvalidArgumentsException::class)
	override fun parseValue(data: ArgumentParsingData): Pair<*, *> {
		return try {
			Pair<Any?, Any?>(typeArgument1.parseValue(data), null)
		} catch (e: InvalidArgumentsException) {
			Pair<Any?, Any?>(null, typeArgument2.parseValue(data))
		}
	}

}